circuits/cpp/src/aztec3/circuits/kernel/public/common.hpp

```diff
@@ -0,0 +1,213 @@
+using DummyComposer = aztec3::utils::DummyComposer;
+using aztec3::circuits::root_from_sibling_path;
+using aztec3::utils::array_length;
+using aztec3::utils::array_pop;
+using aztec3::utils::array_push;
+using aztec3::utils::push_array_to_array;
+
+namespace aztec3::circuits::kernel::public_kernel {
+
+template <typename NT, size_t SIZE>
+void check_membership(DummyComposer& composer,
+                      typename NT::fr const& value,
+                      typename NT::fr const& index,
+                      std::array<typename NT::fr, SIZE> const& sibling_path,
+                      typename NT::fr const& root)
+{
+    const auto calculated_root = root_from_sibling_path<NT>(value, index, sibling_path);
+    composer.do_assert(calculated_root == root, "Membership check failed");
+}
+
+NT::fr hash_public_data_tree_value(NT::fr const& value);
+NT::fr hash_public_data_tree_index(NT::fr const& contract_address, NT::fr const& storage_slot);
+
+template <typename NT, template <class> typename KernelInput>
+void validate_state_reads(DummyComposer& composer, KernelInput<NT> const& public_kernel_inputs)
+{
+    // Validates that state reads correspond to the provided membership witnesses
+    const auto& reads = public_kernel_inputs.public_call.public_call_data.call_stack_item.public_inputs.state_reads;
+    const auto& contract_address = public_kernel_inputs.public_call.public_call_data.call_stack_item.contract_address;
+    for (size_t i = 0; i < STATE_READS_LENGTH; ++i) {
+        const auto& state_read = reads[i];
+        if (state_read.is_empty()) {
+            continue;
+        }
+        const auto& sibling_path = public_kernel_inputs.public_call.state_reads_sibling_paths[i].sibling_path;
+        const typename NT::fr leaf_value = hash_public_data_tree_value(state_read.current_value);
+        const typename NT::fr leaf_index = hash_public_data_tree_index(contract_address, state_read.storage_slot);
+        check_membership<NT>(
+            composer, leaf_value, leaf_index, sibling_path, public_kernel_inputs.public_call.public_data_tree_root);
+    }
+};
+
+template <typename NT, template <class> typename KernelInput>
+void validate_state_transitions(DummyComposer& composer, KernelInput<NT> const& public_kernel_inputs)
+{
+    // Validates that the old value of state transitions correspond to the provided membership witnesses
+    const auto& transitions =
+        public_kernel_inputs.public_call.public_call_data.call_stack_item.public_inputs.state_transitions;
+    const auto& contract_address = public_kernel_inputs.public_call.public_call_data.call_stack_item.contract_address;
+    for (size_t i = 0; i < STATE_TRANSITIONS_LENGTH; ++i) {
+        const auto& state_transition = transitions[i];
+        if (state_transition.is_empty()) {
+            continue;
+        }
+        const auto& sibling_path = public_kernel_inputs.public_call.state_reads_sibling_paths[i].sibling_path;
```

Comment:
Should this be `state_transitions_sibling_paths`?

Reply:
Yes, good catch.